routes/public.js

```diff
@@ -59,6 +59,47 @@ function PublicRouter(app) {
         });
     })
 
+    router.get('/auth/google', function(req, res, next) {
+        passport.authenticate('google', { scope: ['https://www.googleapis.com/auth/plus.login'] }, function(err, user, info) {
+            if (!user) return responseFactory.sendRenderedResponse("public/signin", req, res, { error: info.error || { message: "An unknown error occurred." }, username: req.body.username });
+            req.login(user, function(err) {
+                if (err) return responseFactory.sendRenderedResponse("public/signin", req, res, { error: { message: "An unknown error occurred." }, username: req.body.username });
+                return res.redirect("/?signedin=1");
+            });
+        })(req, res, next);
+    })
+
+    router.get('/auth/google/callback', passport.authenticate('google', { failureRedirect: '/signup' }), function(req, res) {
+        res.redirect('/?signedin=1');
+    })
+
+    router.get('/auth/discord', passport.authenticate('discord'));
+    router.get('/auth/discord/callback', passport.authenticate('discord', {
+        failureRedirect: '/signup',
```

Comment:
Maybe instead of copying and pasting this almost identical code three times it could be in a common function?